src/download-url.ts

```diff
@@ -0,0 +1,125 @@
+import { compareVersions, satisfies, validate } from "compare-versions";
+import { Input } from "./action";
+import { getArchitecture, getPlatform, request } from "./utils";
+
+export interface DownloadMeta {
+  url: string;
+  auth?: string;
+}
+
+export async function getDownloadMeta(options: Input): Promise<DownloadMeta> {
+  const { customUrl } = options;
+  if (customUrl) {
+    return {
+      url: customUrl,
+    };
+  }
+
+  if (options.version && /^[0-9a-f]{40}$/i.test(options.version)) {
+    return await getShaDownloadMeta(options);
+  }
+
+  return await getSemverDownloadMeta(options);
+}
+
+interface Run {
+  id: string;
+  head_sha: string;
+}
+
+interface Runs {
+  workflow_runs: Run[];
+}
+
+async function getShaDownloadMeta(options: Input): Promise<DownloadMeta> {
+  let res: Runs;
+  let page = 1;
+  let run: Run;
+  while (
+    (res = (await (
+      await request(
+        `https://api.github.com/repos/oven-sh/bun/actions/workflows/ci.yml/runs?per_page=100&page=${page}`,
+        {
+          headers: {
+            "Authorization": `Bearer ${options.token}`,
+          },
+        }
+      )
+    ).json()) as Runs)
+  ) {
+    run = res.workflow_runs.find((item) => item.head_sha === options.version);
+    if (run) break;
+
+    page++;
+  }
+
+  const artifacts = (await (
+    await request(
+      `https://api.github.com/repos/oven-sh/bun/actions/runs/${run.id}/artifacts`,
+      {
+        headers: {
+          "Authorization": `Bearer ${options.token}`,
```

Comment:
What if the token is not defined in the configuration? The `headers` field add could be make conditional